init-pants/action.yaml

```diff
@@ -0,0 +1,98 @@
+---
+name: Initialize Pantsbuild Environment
+description: |
+  Install pants and its dependencies and prepare pants caches.
+
+inputs:
+  pants-python-version:
+    required: false
+    default: '3.9'
+  base-branch:
+    required: false
+    default: main
+  gha-cache-key:
+    required: true
+    # use this for matrix.python-version or similar
+  pants-ci-config:
+    required: false
+    default: pants.ci.toml
+    # set to empty string to disable
+
+runs:
+  using: "composite"
+  steps:
+    - name: 'Set up Python for Pants (${{ inputs.pants-python-version }})'
+      uses: actions/setup-python@v2
+      with:
+        python-version: '${{ inputs.pants-python-version }}'
+
+    - name: Get Pants version
+      id: pants_version
+      shell: bash
+      run: |
+        # Capture the "pants_version = " line from config.
+        PANTS_VERSION=$(grep -r '^pants_version\s*=' pants.toml)
+        echo "::set-output name=pants_version::$PANTS_VERSION"
+
+    - name: Cache Pants Setup (${{ steps.pants_version.outputs.pants_version }}-py${{ inputs.pants-python-version }})
+      id: cache-pants-setup
+      uses: actions/cache@v2
+      with:
+        path: |
+          ~/.cache/pants/setup
+        key: pants-setup-${{ runner.os }}-${{ steps.pants_version.outputs.pants_version }}-py${{ inputs.pants-python-version }}
+
+    # Looking up the commit allows us to use the cache
+    # from the latest commit on the base branch.
+    - name: Get Pants Cache Commit (base branch commit to pull cache from)
+      id: pants-cache-commit
+      shell: bash
+      # we could use this, but only if fetch-depth goes back far enough
+      # COMMIT=$(git merge-base ${GITHUB_BASE_REF:-${{ inputs.base-branch }}} HEAD | head -n1)
+      run: |
+        CACHECOMMIT=$(
+          gh api -H "Accept: application/vnd.github+json" \
+          '/repos/${{ github.repository }}/commits?per_page=1&sha=${{ inputs.base-branch }}' \
+          -q '.[].sha'
+        )
+        echo "::set-output name=CACHECOMMIT::${CACHECOMMIT}"
+      env:
+        GH_TOKEN: ${{ github.token }}
+
+    - name: Cache Pants Named Caches (${{ inputs.gha-cache-key }})
+      uses: actions/cache@v2
+      id: cache-pants-named-caches
+      with:
+        path: |
+          ~/.cache/pants/named_caches
+        key: pants-named-caches-${{ runner.os }}-${{ inputs.gha-cache-key }}-${{ github.sha }}
```

Comment:
Named caches is a generic mechanism, but today is primarily used to cache wheels for all transitive resolved requirements. So it can get large pretty quickly. However it only changes when requirements.txt changes. Do we know if the 10GB GHA cache size limit applies to cache values deduped across multiple keys, or naively? If the latter, this will blow that limit pretty quickly and become useless.

A more precise key here would be to hash all the requirements.txt files and any BUILD files that have python_requirement/python_requirements targets.

Reply:
Any sane way to get the list of requirements files before pants has been bootstrapped? I suppose we could make it an input parameter that takes a list of files to hash.

Maybe I should use the sha only for the lmdb store cache. That's what I was thinking of when I went with sha.

Reply:
Input parameters have to be a string, so I would probably have to use a comma separated list of files, or better yet a JSON string.

Reply:
Ooh. I've got it. We should hash the lockfile(s).
It should still be an input parameter, but I am satisfied with how clean using the lockfile should be. I didn't want to hash BUILD files.

Also, the GHA cache uses a tar, zip, or similar archive file, so, files are not shared between cache keys.

Reply:
K. I think I've addressed this, wdyt?